File: src/components/SalesAnalysis.py
import dash
from dash import dcc, html
from dash.dependencies import Input, Output
import plotly.graph_objects as go
import numpy as np
import pandas as pd
from datetime import datetime, timedelta

# Initialize the Dash app
app = dash.Dash(__name__, suppress_callback_exceptions=True)

# Generate sample data
def generate_sample_data():
    np.random.seed(42)
    dates = pd.date_range(start='2024-01-01', end='2024-03-31', freq='D')
    sales = np.random.normal(1000, 200, len(dates)) + np.linspace(0, 500, len(dates))
    return pd.DataFrame({
        'Date': dates,
        'Sales': sales,
        'Units': np.random.randint(50, 150, len(dates)),
        'Profit': sales * np.random.uniform(0.2, 0.4, len(dates))
    })

# Create the layout
layout = html.Div(
    style={'padding': '20px', 'backgroundColor': '#f8f9fa', 'minHeight': '100vh'},
    children=[
        html.Div(
            style={'padding': '20px', 'margin': '10px', 'borderRadius': '12px', 
                  'boxShadow': '0 4px 6px rgba(0, 0, 0, 0.1)',
                  'background': 'linear-gradient(135deg, #ffffff 0%, #f8f9fa 100%)'},
            children=[
                html.H1('Sales Analysis Dashboard', 
                       style={'color': '#1a237e', 'fontWeight': '600', 
                             'textAlign': 'center', 'marginBottom': '20px'}),
                html.Div(
                    style={'display': 'flex', 'gap': '20px', 'justifyContent': 'center'},
                    children=[
                        dcc.Dropdown(
                            id='sales-metric-dropdown',
                            options=[
                                {'label': 'Sales Revenue', 'value': 'Sales'},
                                {'label': 'Units Sold', 'value': 'Units'},
                                {'label': 'Profit', 'value': 'Profit'}
                            ],
                            value='Sales',
                            style={'width': '200px', 'backgroundColor': '#ffffff'}
                        ),
                        dcc.Dropdown(
                            id='sales-timeframe-dropdown',
                            options=[
                                {'label': 'Daily', 'value': 'D'},
                                {'label': 'Weekly', 'value': 'W'},
                                {'label': 'Monthly', 'value': 'M'}
                            ],
                            value='D',
                            style={'width': '200px', 'backgroundColor': '#ffffff'}
                        )
                    ]
                )
            ]
        ),
        
        html.Div(
            style={'display': 'flex', 'flexWrap': 'wrap', 'gap': '20px', 'padding': '20px'},
            children=[
                # Sales Trend
                html.Div(
                    style={'flex': '1', 'minWidth': '45%', 'padding': '20px', 'margin': '10px',
                          'borderRadius': '12px', 'boxShadow': '0 4px 6px rgba(0, 0, 0, 0.1)',
                          'background': 'linear-gradient(135deg, #ffffff 0%, #f8f9fa 100%)'},
                    children=[
                        html.H2('Sales Trend', 
                               style={'color': '#1a237e', 'fontWeight': '600',
                                     'marginBottom': '20px'}),
                        dcc.Graph(id='sales-trend-graph',
                                style={'height': '400px'})
                    ]
                ),
                
                # Distribution
                html.Div(
                    style={'flex': '1', 'minWidth': '45%', 'padding': '20px', 'margin': '10px',
                          'borderRadius': '12px', 'boxShadow': '0 4px 6px rgba(0, 0, 0, 0.1)',
                          'background': 'linear-gradient(135deg, #ffffff 0%, #f8f9fa 100%)'},
                    children=[
                        html.H2('Distribution', 
                               style={'color': '#1a237e', 'fontWeight': '600',
                                     'marginBottom': '20px'}),
                        dcc.Graph(id='sales-distribution-graph',
                                style={'height': '400px'})
                    ]
                ),
                
                # Summary Statistics
                html.Div(
                    style={'flex': '1', 'minWidth': '45%', 'padding': '20px', 'margin': '10px',
                          'borderRadius': '12px', 'boxShadow': '0 4px 6px rgba(0, 0, 0, 0.1)',
                          'background': 'linear-gradient(135deg, #ffffff 0%, #f8f9fa 100%)'},
                    children=[
                        html.H2('Summary Statistics', 
                               style={'color': '#1a237e', 'fontWeight': '600',
                                     'marginBottom': '20px'}),
                        html.Div(id='sales-summary-stats',
                                style={'display': 'flex', 'flexWrap': 'wrap', 'gap': '20px'})
                    ]
                )
            ]
        )
    ]
)

# Set the app's layout
app.layout = layout

# Callbacks
@app.callback(
    [Output('sales-trend-graph', 'figure'),
     Output('sales-distribution-graph', 'figure'),
     Output('sales-summary-stats', 'children')],
    [Input('sales-metric-dropdown', 'value'),
     Input('sales-timeframe-dropdown', 'value')]
)
def update_graphs(selected_metric, selected_timeframe):
    df = generate_sample_data()
    
    if selected_timeframe != 'D':
        df = df.resample(selected_timeframe, on='Date').mean()
    
    # Update trend graph
    trend_fig = go.Figure()
    trend_fig.add_trace(go.Scatter(
        x=df.index if selected_timeframe != 'D' else df['Date'],
        y=df[selected_metric],
        mode='lines+markers',
        name=selected_metric,
        line=dict(color='#3498db', width=2),
        marker=dict(size=8, color='#1a237e')
    ))
    trend_fig.update_layout(
        title=f'{selected_metric} Trend',
        xaxis_title='Date',
        yaxis_title=selected_metric,
        plot_bgcolor='rgba(0,0,0,0)',
        paper_bgcolor='rgba(0,0,0,0)',
        font=dict(color='#1a237e'),
        showlegend=True,
        legend=dict(
            orientation="h",
            yanchor="bottom",
            y=1.02,
            xanchor="right",
            x=1
        )
    )
    
    # Update distribution graph
    dist_fig = go.Figure()
    dist_fig.add_trace(go.Histogram(
        x=df[selected_metric],
        nbinsx=30,
        marker_color='#3498db',
        opacity=0.7
    ))
    dist_fig.update_layout(
        title=f'{selected_metric} Distribution',
        xaxis_title=selected_metric,
        yaxis_title='Frequency',
        plot_bgcolor='rgba(0,0,0,0)',
        paper_bgcolor='rgba(0,0,0,0)',
        font=dict(color='#1a237e'),
        showlegend=False
    )
    
    # Update summary statistics
    stats = df[selected_metric].describe()
    summary_stats = [
        html.Div(
            style={'flex': '1', 'minWidth': '200px', 'padding': '20px',
                  'backgroundColor': '#ffffff', 'borderRadius': '8px',
                  'boxShadow': '0 2px 4px rgba(0,0,0,0.1)'},
            children=[
                html.H3(name.title(), style={'color': '#1a237e', 'marginBottom': '10px'}),
                html.P(f"{value:,.2f}", style={'fontSize': '24px', 'color': '#3498db'})
            ]
        )
        for name, value in stats.items()
    ]
    
    return trend_fig, dist_fig, summary_stats

if __name__ == '__main__':
    app.run(debug=True) 

File: src/components/SalesAnalysis.js
import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Grid, Paper, Typography, Select, MenuItem, FormControl, InputLabel, Button } from '@material-ui/core';
import Plot from 'react-plotly.js';
import axios from 'axios';

const useStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(3),
    background: '#ffffff',
    minHeight: '100vh',
  },
  paper: {
    padding: theme.spacing(3),
    margin: theme.spacing(2),
    borderRadius: '12px',
    boxShadow: '0 4px 6px rgba(0, 0, 0, 0.1)',
  },
  title: {
    marginBottom: theme.spacing(2),
    color: '#1a237e',
    fontWeight: 600,
  },
  formControl: {
    margin: theme.spacing(1),
    minWidth: 200,
  },
  description: {
    marginTop: theme.spacing(2),
    color: '#666',
    lineHeight: 1.6,
  },
  graphContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'flex-start',
    gap: theme.spacing(3),
  },
  graph: {
    flex: 1,
  },
  textContainer: {
    flex: 1,
    padding: theme.spacing(2),
  },
}));

function SalesAnalysis() {
  const classes = useStyles();
  const [selectedMetric, setSelectedMetric] = useState('revenue');
  const [timeFrame, setTimeFrame] = useState('monthly');
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    // Generate sample data immediately
    const sampleData = Array.from({ length: 30 }, (_, i) => ({
      date: new Date(2023, 0, i + 1),
      revenue: Math.floor(Math.random() * 10000) + 1000,
      units: Math.floor(Math.random() * 100) + 10,
      profit: Math.floor(Math.random() * 5000) + 500,
      customers: Math.floor(Math.random() * 50) + 5,
    }));
    setData(sampleData);
  }, []);

  const handleMetricChange = (event) => {
    setSelectedMetric(event.target.value);
  };

  const handleTimeFrameChange = (event) => {
    setTimeFrame(event.target.value);
  };

  return (
    <div className={classes.root}>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <Paper className={classes.paper}>
            <Typography variant="h5" className={classes.title}>
              Sales Analysis Dashboard
            </Typography>
            <Grid container spacing={2}>
              <Grid item xs={12} md={6}>
                <FormControl className={classes.formControl}>
                  <InputLabel>Metric</InputLabel>
                  <Select
                    value={selectedMetric}
                    onChange={handleMetricChange}
                  >
                    <MenuItem value="revenue">Revenue</MenuItem>
                    <MenuItem value="units">Units Sold</MenuItem>
                    <MenuItem value="profit">Profit</MenuItem>
                    <MenuItem value="customers">Customers</MenuItem>
                  </Select>
                </FormControl>
              </Grid>
              <Grid item xs={12} md={6}>
                <FormControl className={classes.formControl}>
                  <InputLabel>Time Frame</InputLabel>
                  <Select
                    value={timeFrame}
                    onChange={handleTimeFrameChange}
                  >
                    <MenuItem value="daily">Daily</MenuItem>
                    <MenuItem value="weekly">Weekly</MenuItem>
                    <MenuItem value="monthly">Monthly</MenuItem>
                  </Select>
                </FormControl>
              </Grid>
            </Grid>
          </Paper>
        </Grid>

        {data && (
          <Grid item xs={12} md={6}>
            <Paper className={classes.paper}>
              <Typography variant="h6" className={classes.title}>
                Sales Trend
              </Typography>
              <div className={classes.graphContainer}>
                <div className={classes.graph} style={{ height: 400 }}>
                  <Plot
                    data={[
                      {
                        x: data.map(d => d.date),
                        y: data.map(d => d[selectedMetric]),
                        type: 'scatter',
                        mode: 'lines+markers',
                        name: selectedMetric,
                        line: { color: '#3498db' },
                        marker: { size: 8 }
                      }
                    ]}
                    layout={{
                      title: `${selectedMetric} Over Time`,
                      xaxis: { title: 'Date' },
                      yaxis: { title: selectedMetric }
                    }}
                  />
                </div>
                <div className={classes.textContainer}>
                  <Typography variant="body1" className={classes.description}>
                    This line chart shows the trend of {selectedMetric} over time. The blue line represents the overall trend, while the markers show individual data points. Use this visualization to:
                  </Typography>
                  <ul>
                    <li>Identify patterns and trends in your sales data</li>
                    <li>Spot seasonal variations or anomalies</li>
                    <li>Track growth or decline over time</li>
                    <li>Compare performance across different time periods</li>
                  </ul>
                </div>
              </div>
            </Paper>
          </Grid>
        )}

        {data && (
          <Grid item xs={12} md={6}>
            <Paper className={classes.paper}>
              <Typography variant="h6" className={classes.title}>
                Distribution
              </Typography>
              <div className={classes.graphContainer}>
                <div className={classes.graph} style={{ height: 400 }}>
                  <Plot
                    data={[
                      {
                        x: data.map(d => d[selectedMetric]),
                        type: 'histogram',
                        name: selectedMetric,
                        marker: { color: '#3498db' }
                      }
                    ]}
                    layout={{
                      title: `${selectedMetric} Distribution`,
                      xaxis: { title: selectedMetric },
                      yaxis: { title: 'Frequency' }
                    }}
                  />
                </div>
                <div className={classes.textContainer}>
                  <Typography variant="body1" className={classes.description}>
                    This histogram shows how your {selectedMetric} is distributed across different value ranges. The height of each bar represents how frequently values fall within that range. Use this visualization to:
                  </Typography>
                  <ul>
                    <li>Understand the typical range of your sales metrics</li>
                    <li>Identify common and rare values</li>
                    <li>Spot outliers or unusual patterns</li>
                    <li>Assess the spread and concentration of your data</li>
                  </ul>
                </div>
              </div>
            </Paper>
          </Grid>
        )}

        {data && (
          <Grid item xs={12}>
            <Paper className={classes.paper}>
              <Typography variant="h6" className={classes.title}>
                Summary Statistics
              </Typography>
              <Grid container spacing={2}>
                <Grid item xs={12} md={3}>
                  <Typography variant="h6">Total {selectedMetric}</Typography>
                  <Typography variant="h4">
                    {data.reduce((sum, d) => sum + d[selectedMetric], 0).toLocaleString()}
                  </Typography>
                  <Typography variant="body2" color="textSecondary">
                    The sum of all {selectedMetric} values in the selected period
                  </Typography>
                </Grid>
                <Grid item xs={12} md={3}>
                  <Typography variant="h6">Average {selectedMetric}</Typography>
                  <Typography variant="h4">
                    {(data.reduce((sum, d) => sum + d[selectedMetric], 0) / data.length).toLocaleString()}
                  </Typography>
                  <Typography variant="body2" color="textSecondary">
                    The mean value of {selectedMetric} per time period
                  </Typography>
                </Grid>
                <Grid item xs={12} md={3}>
                  <Typography variant="h6">Max {selectedMetric}</Typography>
                  <Typography variant="h4">
                    {Math.max(...data.map(d => d[selectedMetric])).toLocaleString()}
                  </Typography>
                  <Typography variant="body2" color="textSecondary">
                    The highest value of {selectedMetric} achieved
                  </Typography>
                </Grid>
                <Grid item xs={12} md={3}>
                  <Typography variant="h6">Min {selectedMetric}</Typography>
                  <Typography variant="h4">
                    {Math.min(...data.map(d => d[selectedMetric])).toLocaleString()}
                  </Typography>
                  <Typography variant="body2" color="textSecondary">
                    The lowest value of {selectedMetric} recorded
                  </Typography>
                </Grid>
              </Grid>
            </Paper>
          </Grid>
        )}
      </Grid>
    </div>
  );
}

export default SalesAnalysis; 